src/darker/formatters/black_formatter.py

```diff
@@ -0,0 +1,205 @@
+            if isinstance(target_version, str):
+                self.config["target_version"] = target_version
+            elif isinstance(target_version, list):
+                # Convert TOML list to a Python set
+                self.config["target_version"] = set(target_version)
+            else:
+                message = (
+                    f"Invalid target-version = {target_version!r} in {config_path}"
+                )
+                raise ConfigurationError(message)
+        if "exclude" in raw_config:
+            self.config["exclude"] = re_compile_maybe_verbose(raw_config["exclude"])
+        if "extend_exclude" in raw_config:
+            self.config["extend_exclude"] = re_compile_maybe_verbose(
+                raw_config["extend_exclude"]
+            )
+        if "force_exclude" in raw_config:
+            self.config["force_exclude"] = re_compile_maybe_verbose(
+                raw_config["force_exclude"]
+            )
+
+    def _read_cli_args(self, args: Namespace) -> None:
+        return read_black_compatible_cli_args(args, self.config)
+
+    def run(self, content: TextDocument, path_from_cwd: Path) -> TextDocument:
+        """Run the Black code re-formatter for the Python source code given as a string.
+
+        :param content: The source code
+        :param path_from_cwd: The path to the source code file being reformatted, either
+                              absolute or relative to the current working directory
+        :return: The reformatted content
+
+        """
+        # Collect relevant Black configuration options from ``self.config`` in order to
+        # pass them to Black's ``format_str()``. File exclusion options aren't needed
+        # since at this point we already have a single file's content to work on.
+        mode = BlackModeAttributes()
+        if "line_length" in self.config:
+            mode["line_length"] = self.config["line_length"]
+        if "target_version" in self.config:
+            all_target_versions = {tgt_v.name.lower(): tgt_v for tgt_v in TargetVersion}
+            target_versions_in = validate_target_versions(
+                self.config["target_version"], all_target_versions
+            )
+            mode["target_versions"] = {
+                all_target_versions[n] for n in target_versions_in
+            }
+        if "skip_magic_trailing_comma" in self.config:
+            mode["magic_trailing_comma"] = not self.config["skip_magic_trailing_comma"]
+        if "skip_string_normalization" in self.config:
+            # The ``black`` command line argument is
+            # ``--skip-string-normalization``, but the parameter for
+            # ``black.Mode`` needs to be the opposite boolean of
+            # ``skip-string-normalization``, hence the inverse boolean
+            mode["string_normalization"] = not self.config["skip_string_normalization"]
+        if "preview" in self.config:
+            mode["preview"] = self.config["preview"]
```

Comment:
IMO above should be moved out of run function since just related to setting up config of the class

Reply:
Setting up the config of the formatter class is done in `_read_config_file()`.

This code is about preparing the options for the call to `black.format_str(contents_for_black, mode=Mode(**mode))` which are encapsulated inside the `Mode` object.

But true nevertheless – preparing the data for the `Mode` object can indeed be split into a helper method.